File: source/sindri/sindri/saver/at_perturbation.py
# encoding: utf-8

# Copyright (c) 2001-2014, Canal TP and/or its affiliates. All rights reserved.
#
# This file is part of Navitia,
#     the software to build cool stuff with public transport.
#
# Hope you'll enjoy and contribute to this project,
#     powered by Canal TP (www.canaltp.fr).
# Help us simplify mobility and open public transport:
#     a non ending quest to the responsive locomotion way of traveling!
#
# LICENCE: This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# Stay tuned using
# twitter @navitia
# IRC #navitia on freenode
# https://groups.google.com/d/forum/navitia
# www.navitia.io

from sqlalchemy import select
from sindri.saver.utils import parse_active_days, from_timestamp, from_time, \
    FunctionalError


def build_at_perturbation_dict(perturbation):
    """
    construit à partir d'un object protobuf pbnavitia.realtime.AtPerturbation
    le dictionnaire utilisé pour l'insertion en base
    """
    result = {}
    result['uri'] = perturbation.uri
    result['object_uri'] = perturbation.object.object_uri
    result['object_type_id'] = perturbation.object.object_type

    result['active_days'] = parse_active_days(perturbation.active_days)

    result['start_application_date'] = from_timestamp(
        perturbation.start_application_date)
    result['end_application_date'] = from_timestamp(
        perturbation.end_application_date)

    result['start_application_daily_hour'] = from_time(
        perturbation.start_application_daily_hour)
    result['end_application_daily_hour'] = from_time(
        perturbation.end_application_daily_hour)

    return result


def persist_at_perturbation(meta, conn, perturbation):
    """
    enregistre en base la perturbation at
    """
    row_id = find_at_perturbation_id(meta, conn, perturbation.uri)
    save_at_perturbation(meta, conn, row_id, perturbation)


def find_at_perturbation_id(meta, conn, perturbation_uri):
    """
    retourne l'id en base de la perturbation correspondant à cette URI
    si celle ci est présente dans ED sinon retourne None
    """
    perturbation_table = meta.tables['realtime.at_perturbation']
    query = select([perturbation_table.c.id],
                   perturbation_table.c.uri == perturbation_uri)
    result = conn.execute(query)
    row = result.fetchone()
    return row[0] if row else None


def save_at_perturbation(meta, conn, perturbation_id, perturbation):
    """
    update ou insert la perturbation
    """
    perturbation_table = meta.tables['realtime.at_perturbation']
    if not perturbation_id:
        query = perturbation_table.insert()
    else:
        query = perturbation_table.update().where(
            perturbation_table.c.id == perturbation_id)

    conn.execute(query.values(build_at_perturbation_dict(
        perturbation)))

File: source/ed/at_perturbation.cpp
/* Copyright © 2001-2014, Canal TP and/or its affiliates. All rights reserved.

This file is part of Navitia,
    the software to build cool stuff with public transport.

Hope you'll enjoy and contribute to this project,
    powered by Canal TP (www.canaltp.fr).
Help us simplify mobility and open public transport:
    a non ending quest to the responsive locomotion way of traveling!

LICENCE: This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

Stay tuned using
twitter @navitia
IRC #navitia on freenode
https://groups.google.com/d/forum/navitia
www.navitia.io
*/
#include "at_perturbation.h"
#include <boost/date_time/gregorian/gregorian.hpp>

namespace pt = boost::posix_time;
namespace bg = boost::gregorian;

namespace ed {

bool AtPerturbation::is_applicable(const boost::posix_time::time_period& period) const{
    bool days_intersects = false;

    //intersection de la period ou se déroule l'action et de la periode de validité de la perturbation
    pt::time_period common_period = period.intersection(application_period);

    //si la periode commune est null, l'impact n'est pas valide
    if(common_period.is_null()){
        return false;
    }

    //on doit travailler jour par jour
    //
    bg::date current_date = common_period.begin().date();

    while(!days_intersects && current_date <= common_period.end().date()){
        //on test uniquement le debut de la period, si la fin est valide, elle sera testé à la prochaine itération
        days_intersects = valid_day_of_week(current_date);

        //vérification des plages horaires journaliéres
        pt::time_period current_period = common_period.intersection(pt::time_period(pt::ptime(current_date, pt::seconds(0)), pt::hours(24)));
        days_intersects = days_intersects && valid_hour_perturbation(current_period);

        current_date += bg::days(1);
    }

    return days_intersects;
}

bool AtPerturbation::valid_hour_perturbation(const pt::time_period& period) const{
    pt::time_period daily_period(pt::ptime(period.begin().date(), application_daily_start_hour),
            pt::ptime(period.begin().date(), application_daily_end_hour));

    return period.intersects(daily_period);
}

bool AtPerturbation::valid_day_of_week(const bg::date& date) const{

    switch(date.day_of_week()){
        case bg::Monday: return this->active_days[0];
        case bg::Tuesday: return this->active_days[1];
        case bg::Wednesday: return this->active_days[2];
        case bg::Thursday: return this->active_days[3];
        case bg::Friday: return this->active_days[4];
        case bg::Saturday: return this->active_days[5];
        case bg::Sunday: return this->active_days[6];
    }
    return false; // Ne devrait pas arriver
}
}
